### PR TITLE
Assignment 6

### PR DESCRIPTION
**Pull request by crackerJackTech**

### CODE REVIEW

File: Array.Tests/Arrays.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+
```
Comment (samrnixon) [inline Array.Tests/Arrays.Tests.csproj]:
```suggestion
<nullable>enable</nullable>
```

File: Array.Tests/Arrays.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
```
Comment (MarkMichaelis) [inline Array.Tests/Arrays.Tests.csproj]: Why are there two Array test projects?

File: Assignment6/Array.cs [hunk 3]
```diff
@@ -1,10 +1,117 @@
...
+using System.Collections;
 using System.Collections.Generic;
 
 namespace Assignment6
 {
-    public class Array<T>
+    public class ArrayCollection<T> : ICollection<T>
     {
+        private List<T> List { get; }
+        public int Capacity { get; }
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Glad to see this is read-only.

File: Assignment6/Array.cs [hunk 4]
```diff
@@ -1,10 +1,117 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
 namespace Assignment6
 {
-    public class Array<T>
+    public class ArrayCollection<T> : ICollection<T>
     {
+        private List<T> List { get; }
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: This makes the most sense for what type of collection to use internally and I am glad to see it is private.

File: Assignment6/Array.cs [hunk 5]
```diff
@@ -1,10 +1,117 @@
...
+            {
+                throw new ArgumentException("That item doesn't exist in the Array", nameof(item));
+            }
+
+            return List.Contains(item);
+        }
+
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            if(arrayIndex >= Capacity)
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: What if the `arrayIndex` was negative?

File: Assignment6/Array.cs [hunk 6]
```diff
@@ -1,10 +1,117 @@
...
+            if(List.Remove(item))
+            {
+                NumberOfItemsInArray--;
+                return true;
+            }
+
+            return false;
+        }
+
+        public IEnumerator<T> GetEnumerator()
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Consider using given that this implementation is so simple:
```
public IEnumerator<T> GetEnumerator() => List.GetEnumerator();
```

File: Assignment6/Array.cs [hunk 7]
```diff
@@ -1,10 +1,117 @@
...
+
+        public void Clear()
+        {
+            List.Clear();
+            NumberOfItemsInArray = 0;
+        }
+
+        public bool Contains(T item)
+        {
+            if(item is null)
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: It is not clear to me why you don't allow null to be stored?

File: Assignment6/Array.cs [hunk 8]
```diff
@@ -1,10 +1,117 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
 namespace Assignment6
 {
-    public class Array<T>
+    public class ArrayCollection<T> : ICollection<T>
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Class name `ArrayCollection` does not match file name, `Array.cs`.

File: Assignment6/Array.cs [hunk 9]
```diff
@@ -1,10 +1,117 @@
...
+                return List[i];
+            }
+
+            set
+            {
+                List.Add(List[i]);
+            }
+        }
+
+        public void Add(T item)
```
Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Supporting Add and Remove are not typical array capabilities.  I would not have these implementations.

Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Supporting Add and Remove are not typical array capabilities.  I would not have these methods.

Comment (MarkMichaelis) [inline Assignment6/Array.cs]: Supporting Add and Remove are not typical array capabilities.  I would not have these methods except as explicit implementations of ICollection<T>.

File: Arrays.Tests/ArraysTests.cs [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+            Assert.AreEqual(test3.Capacity, size);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentOutOfRangeException))]
+        public void Constructor_InitializedArrayWithMoreDataThanSize_ThrowsException()
+        {
+            //Arrange
+            var sizeOfArray = 0;
+            var _ = new ArrayCollection<string>(sizeOfArray)
```
Comment (MarkMichaelis) [inline Arrays.Tests/ArraysTests.cs]: Not sure you are adding value by including `var` when you are discarding.  Use just `_` IMO.

**Comment by MarkMichaelis** [comment]
##  (crackerJackTech)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), 0 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **17** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ✔ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |

| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **❌** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Arrays.Tests* | Disable |  |  |  | 4 |
| Arrays.Tests | enable| Assignment6.sln |  | CA1303, CA2208, CA1065 | 4 |
| Assignment6 | enable| Assignment6.sln |  | CA1303, CA2208, CA1065 | 4 |
*<sub><sup>The extraneous Array.Tests project should be cleaned up/removed.</sub></sup>


| Testing | **20** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=19, Passed=19, Failed=0, Skipped=0 | **87%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ❌ |

**Pull request closed without merging**